src/server/frontend_wayland/xdg_shell_stable.cpp

```diff
@@ -258,24 +261,19 @@ void mf::XdgSurfaceStable::send_configure()
     send_configure_event(serial);
 }
 
-std::experimental::optional<mf::WindowWlSurfaceRole*> const& mf::XdgSurfaceStable::window_role()
+mw::Weak<mf::WindowWlSurfaceRole> const& mf::XdgSurfaceStable::window_role()
 {
-    if (window_role_ && *window_role_destroyed)
-    {
-        window_role_ = std::experimental::nullopt;
-        window_role_destroyed = nullptr;
-    }
-
     return window_role_;
 }
 
 void mf::XdgSurfaceStable::set_window_role(WindowWlSurfaceRole* role)
 {
-    if (window_role())
+    if (window_role_)
+    {
         log_warning("XdgSurfaceStable::window_role set multiple times");
```

Comment:
Pre-existing, but we ought to provide some information about which client & window is misbehaving.

Reply:
The correct behavior here is to send a Wayland error and kill the client, but knowing what error to send will take a closer look. I believe there may already be a WLCS test for this (at least there are tests for similar things). As you said, pre-existing. I'll look at that another day.